File: common/src/main/java/juuxel/adorn/platform/PlatformBridges.java
package juuxel.adorn.platform;

import juuxel.adorn.util.InlineServices;
import juuxel.adorn.util.Services;

@InlineServices
public interface PlatformBridges {
    BlockEntityBridge getBlockEntities();
    BlockFactory getBlockFactory();
    MenuBridge getMenus();
    NetworkBridge getNetwork();
    ResourceBridge getResources();

    @InlineServices.Getter
    static PlatformBridges get() {
        return Services.load(PlatformBridges.class);
    }
}

File: common/src/main/kotlin/juuxel/adorn/platform/PlatformBridges.kt
package juuxel.adorn.platform

import juuxel.adorn.util.InlineServices
import juuxel.adorn.util.loadService

interface PlatformBridges {
    val blockEntities: BlockEntityBridge
    val blockFactory: BlockFactory
    val entities: EntityBridge
    val menus: MenuBridge
    val network: NetworkBridge
    val resources: ResourceBridge

    @InlineServices
    companion object {
        private val instance: PlatformBridges by lazy { loadService() }
        fun get() = instance

        inline val blockEntities get() = get().blockEntities
        inline val blockFactory get() = get().blockFactory
        inline val entities get() = get().entities
        inline val menus get() = get().menus
        inline val network get() = get().network
        inline val resources get() = get().resources
    }
}
